### PR TITLE
Refactored entity class

### PR DESCRIPTION
**Pull request by Gman98ish**

### CODE REVIEW

File: src/Entity.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    public function toArray($map = [])
+    {
+        $arr = $this->all();
+        foreach ($this->readOnly as $readOnly) {
+            unset($arr[$readOnly]);
+        }
+
+        foreach ($arr as $name => $value) {
+            if ($value instanceof Entity) {
+                unset($arr[$name]);
```
Comment (martinshaw) [inline src/Entity.php]: We are removing the 'sub'-entities here. But later in the client 'toJson' methods we are manually re-appending their JSON representation to the array.

Wouldn't it make sense to just change this line to
```suggestion
                $arr[$name] = $arr[$name]->toArray();
```

And if we don't want to include one of these 'sub-entities' in the array (which we aren't doing in any of the following client classes), we could just manually unset it in the 'toJson' method.

File: src/Entity.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    /**
+     * @var array
+     */
+    protected $attributes = [];
+
+    /**
+     * An array of read only attributes
+     * @var array
+     */
+    protected $readOnly = [];
```
Comment (martinshaw) [inline src/Entity.php]: If we are only using the `readOnly` array to remove columns from the JSON representation. Maybe we could name it `hidden` like its Eloquent equivalent. Just an opinion.

File: src/Entity.php (hunk 1 above)
Comment (Gman98ish) [reply src/Entity.php]: Yeah, good point!

File: src/Entity.php (hunk 2 above)
Comment (Gman98ish) [reply src/Entity.php]: The idea behind this is that we have a list of properties we know not to send on post/patch endpoints. It's similar-ish to eloquent's hidden, but I think this shows its intent more clearly.

File: src/PSS/RequestClient.php [hunk 3]
```diff
@@ -138,80 +138,81 @@ public function getFeedback($id)
      */
```
[the remark this answers is not in the document]
Comment (Gman98ish) [reply src/PSS/RequestClient.php]: :+1:

**Comment by gavtaylor** [comment]
@Gman98ish could you take a look at the conflicts on this, will be good to get this merged in, ta.

File: src/Entity.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+     */
+    public function fill($attributes)
+    {
+        foreach ($attributes as $name => $value) {
+            $this->attributes[$name] = $value;
+        }
+    }
+
+    /**
+     * Returns true if the attribute is present on the
```
Comment (MoHangingWithLegends) [inline src/Entity.php]: Are we checking the existence of the key or value.

File: src/Entity.php (hunk 2 above)
Comment (MoHangingWithLegends) [reply src/Entity.php]: Is it possible to add an array which restricts what can be filled.

File: src/PSS/Entities/Request.php [hunk 5]
```diff
@@ -2,92 +2,32 @@
...
-    /**
-     * @var array
-     */
-    public $cc = [];
-
-    /**
-     * @var integer
-     */
-    public $unreadReplies;
+    protected $readOnly = [
```
Comment (MoHangingWithLegends) [inline src/PSS/Entities/Request.php]: Should the apio not handle this.

File: src/Entity.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    /**
+     * Gets an attribute, if the attribute hasn't been
+     * set yet, return $default
+     *
+     * @param string $attr
+     * @param mixed $default
+     * @return mixed
+     */
+    public function get($attr, $default = null)
```
[the remark this answers is not in the document]
Comment (Gman98ish) [reply src/Entity.php]: I'm not sure that would be clearer, we're not getting a key from the entity, we're getting an attribute

File: src/Entity.php (hunk 2 above)
Comment (Gman98ish) [reply src/Entity.php]: I'm happy to to an array which restricts what can be filled, but I think that'd be better off in a separate PR to keep this one smaller

File: src/Entity.php (hunk 4 above)
Comment (Gman98ish) [reply src/Entity.php]: It checks if the attribute is present on the Entity. So in this instance, it checks if the key is set

**Comment by Gman98ish** [comment]
> Looks good to me, it doesnt look like theres anything preventing setting random attributes on an entity any longer though, if this is important. I like the toArray() method with mapping on Entity. These changes will definitely help with updates. Not spotted anything that needs changing, and the changes should not break existing implementations.

We can look into preventing setting random attributes by parsing the class doc, I suppose?

**Pull request merged**